Add electrode coloration to grid

CodyCBakerPhD opened the pull request:
@bendichter Adds coloration to the electrode grid based on estimated responsitivity from the spike detection.

Responsitivity calculated as (X - mu) / sigma, where X = trial average of evoked spiking rate in a small window (1 seconds) after stimulus presentation, mu = trial average of pre-stimulus spiking rates estimated over a larger time window (10 seconds), sigma = baseline noise calculated from the standard deviation over trials of spiking rates estimated over the larger pre-stimulus time window.

Produces the following coloration for the fully spike detected session.

![image](https://user-images.githubusercontent.com/51133164/124318114-27642400-db46-11eb-9c00-e5cb1fbd9ca6.png)

The yellow units (positive responsitivity) are the ones with good-looking PSTH's (noisy baseline spiking, sharp tuning to stimulus presentation). The dark blue show up as bad units (very low overall spiking, next to no reaction to stimulus). The units with mid-range teal colors have low baseline activity but some limited response to certain stimulus presentations.

bendichter commented:
@CodyCBakerPhD 

Nice! A few questions:
* Why is the background green?
* Why is the colorbar negative?
* Those time windows seem awfully long. Is that the timescale of the task? Maybe we should allow those to be adjusted

CodyCBakerPhD commented:
1. Open to having the background color adjusted to something that looks better (including blank white) - the prior color scheme (seen [here](https://github.com/catalystneuro/feldman-lab-to-nwb/pull/9)) also didn't use this auto-gradient that plotly has under the `"Viridis"` mode, so we could also change that too if you think a different mode would look better
2. Colorbar is negative because responsitivity (from metric above) is negative for the units that don't look great on the PSTH's. Just means the estimate of the stimulus-evoked spiking rate was less than baseline levels (also related to the fact that it's a shorter time window)
3. These can all be adjusted and passed into the `calculate_response` helper function; would you suggest adding direct widget controls for these?

bendichter commented:
@CodyCBakerPhD 

1. I like Viridis. It's one of [the perceptually uniform colormaps](https://bids.github.io/colormap/). I just don't like the background blending in. White or black would be fine.
2. Is that what we would expect? Usually selective neurons increase their firing rate in response to a stimulus, which would give us positive values. Can you check the PSTHs?
3. Can you look at the timescale of the trials in the experiment and the response times in the PSTHs and set reasonable windows accordingly? I think this may fix #2 as well. Ultimately, I think this might require manual controls in the form of a [BoundedFloatText](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#BoundedFloatText) (sliders are problematic if you don't know the range beforehand.

CodyCBakerPhD commented:
@bendichter 

1. Agreed, fixed now.

> 2. Is that what we would expect? Usually selective neurons increase their firing rate in response to a stimulus, which would give us positive values. Can you check the PSTHs?
2. I would expect this from looking at this particular run of the spike detection. Here are some screenshots of the positive vs. negative responses

![image](https://user-images.githubusercontent.com/51133164/124327156-a3b23380-db55-11eb-9186-ec3c29f725a4.png)

![image](https://user-images.githubusercontent.com/51133164/124327218-b9bff400-db55-11eb-89d4-9381f359109c.png)

I would also expect this to dissipate if any amount of auto-curation is applied in the rapid testing pipeline (optional cell to run, I didn't here just to get the most number of detected channels).

3. The default values of the response calculation are tuned specifically to the timescales of this particular experiment setup, shown also in the screenshots above. That can and definitely will change from experiment to experiment. I definitely agree that BoundedFloatText is the way to go for that and for the time intervals on Raster Grids and Tuning Curves (in dev) tabs as well. I'm actually thinking I can just harness the alignment times on the PSTH side for the pre- and post- stimulus windows. Only box I would have to add somewhere is for the timing of the actual stimulus vs. trial start time (100 ms offset here, hardcoded, and not automatically extractable from the rest of the rapid testing pipeline).

bendichter commented:
@CodyCBakerPhD judging by that PSTH, your values are way off. The pre-stimulus period should be 500 ms and the post-stimulus period should be 200 ms.

CodyCBakerPhD commented:
@bendichter Good catch on that, those time intervals are more appropriate for the size of the trial intervals.

I've now coupled the responsivity parameters with those of the controllers on the PSTH side of the widget.

Unfortunately, when we then set the pre- and post- times that you suggest, we still see similar negative values

![image](https://user-images.githubusercontent.com/51133164/124408296-e4f04200-dd13-11eb-9fa1-1167bd2b7014.png)

I think I know why this is: keep in mind there are 8 total stimuli, and most of these units are showing amplified responses 
 (increases compared baseline spike rate) only to a small number of particular stimuli. Similarly, they show suppressed responses (decreases compared to baseline spike rate) to many of the stimuli they are not tuned to.

A simplified diagram for unimodal response: for stimulus range theta, tuning curve tuned to stimulus theta* in black, and baseline spiking rate (_i.e._, no stimulus presented at all) in red. Our metric of responsitivity as defined would be the integration of the difference between the black and red curves, which if the tuning width is substantially thin would reliably give negative values when averaged over the stimulus space.

![image](https://user-images.githubusercontent.com/51133164/124409447-4fa27d00-dd16-11eb-9666-cf79201b3bd9.png)

So, something I've also done to give more informative values, is I've also coupled the responsitivity to the 'group_by' controller: when you select a categorical variable from the trials table, the responsitivity calculates over each unique value and the final scalar returned is set to the maximum of all of these (_i.e._, the blue dashed region above). With this metric, the returned values are nearly all positive.

![image](https://user-images.githubusercontent.com/51133164/124409152-c428ec00-dd15-11eb-93a9-eebe2206cae1.png)

![image](https://user-images.githubusercontent.com/51133164/124409216-e0c52400-dd15-11eb-88d1-4674e7d71548.png)

bendichter commented:
@CodyCBakerPhD I see what you mean but it looks like the baseline firing rate is at the bottom of the bell curve in this case, so I'm still not understanding why we are getting these results. Would you be able to try using the interval [0.1 0.15] for the stimulus response?

CodyCBakerPhD commented:
@bendichter Even looking at the [0.1, 0.15] evoked period doesn't change much, the most negative units become even more negative and the couple good units stay about the same level

![image](https://user-images.githubusercontent.com/51133164/124501531-b912a200-dd8f-11eb-99ba-5585635b6c20.png)

There may be some confusion from how I was showing one of the only good units (yellow color, has been positive from the beginning of the PR). Here are some of the bad ones, which clearly either (a) have lower evoked rates than pre-stim, or (b) have no data for pre-stim ones.

Example of (a)

![image](https://user-images.githubusercontent.com/51133164/124501297-499cb280-dd8f-11eb-8110-ab914861260a.png)

Example of (b), which is actually like 60% of the units said to be 'detected' (auto-curation likely would have tossed these out, though)

![image](https://user-images.githubusercontent.com/51133164/124501355-60dba000-dd8f-11eb-9d39-ec440bf97686.png)

(b) actually is unresolved at this point: they have `np.nan` responses, and the color shows up as a muted grey with nothing on the colorbar to indicate those points are special in that they don't have sufficient data to calculate a response. What I'll do for that is something like [here](https://community.plotly.com/t/visually-marking-numpys-nan-as-grey/17007/2) where it shows up on the bottom (but with a better tick text label).

Do you have a recommended color for setting the nans to make them distinct from the Viridis colorbar for the valid values?

bendichter commented on feldman_lab_to_nwb/convert_feldman/feldman_electrode_widget.py:
```diff
@@ -1,51 +1,183 @@
+SELECTED_SIZE = 20
+
+
+def calculate_response(
```
@CodyCBakerPhD Could you make the "responsiveness calculation done in a function on a per-unit basis and call it in a loop? I think that way would be easier to test and to follow.

CodyCBakerPhD replied:
@bendichter I've broken up the main code to operate in more relevant pieces on a per-unit basis. The primary `calculate_response` function still includes the iteration over units, but passes on the lifting down to the other helper functions in the new utils file.

I'll go ahead and share the NWBFile I'm working with on Slack as well; to get the full internal variables returned for inspection simply run

```
with NWBHDF5IO(path=nwbfile_name, mode="r") as io:
    nwbfile = io.read()
    response, internals = calculate_response(units=nwbfile.units, trials=nwbfile.trials, pre_alignment_window=[0, 0.5], evoked_window=[0.1, 0.15], event_name="start_time")
```

CodyCBakerPhD replied:
For easy reference, to see it from the widget view with this branch checked out,
```
from feldman_lab_to_nwb import nwb2widget
from pynwb import NWBHDF5IO


nwbfile_name = ".../rapid_testing_LR_210209_g0_full.nwb"

io = NWBHDF5IO(nwbfile_name, "r")
nwbfile=io.read()
nwb2widget(nwbfile)
```
to which I have now added a temporary box for setting the offset for evoked responsitivity calculation. I've also added the raw responsitivity value for the channels as text to the channel hover box on the grid.

CodyCBakerPhD replied:
Temporary box removed, now coupled with PSTH view.

bendichter commented on feldman_lab_to_nwb/convert_feldman/feldman_electrode_widget.py:
```diff
@@ -1,51 +1,65 @@
         pass
 
-    def __init__(self, electrodes):
+    def __init__(self, electrodes, psth_info):
```
It's not clear what "psth_info" should be. Can you please add a docstring?

CodyCBakerPhD replied:
Renamed more appropriately to `responsitivity_params`; the PSTH happens to be where the info is coming right now, but in its current form it's already generalized beyond that.

Added docstring for its form. It also doesn't have to use the dictionary packing if that's adding confusion, I was just doing that for convenience in passing between the widgets.

CodyCBakerPhD replied:
Dictionary compression is no longer used here, just direct keyword arguments.

CodyCBakerPhD commented:
@bendichter With that range issue now fixed with respect to the trial subsetting in the average calculations, I've added docstrings and cleaned up the code.

Widget now produces
![image](https://user-images.githubusercontent.com/51133164/126000899-9ceb8c83-9798-4f26-bccd-bd4967bf53ac.png)

bendichter reviewed (approved):
Looks great!
